lib/commons/dom/is-hidden-with-css.js

```diff
@@ -0,0 +1,50 @@
+/* global dom */
+
+/**
+ * Determine whether an element is hidden based on css
+ * @method isHiddenWithCSS
+ * @memberof axe.commons.dom
+ * @instance
+ * @param {HTMLElement} el The HTML Element
+ * @param {Boolean} descendentVisibilityValue (Optional) immediate descendant visibility value used for recursive computation
+ * @return {Boolean} the element's hidden status
+ */
+dom.isHiddenWithCSS = function isHiddenWithCSS(el, descendentVisibilityValue) {
+	// 9 === Node.DOCUMENT
+	if (el.nodeType === 9) {
+		return false;
+	}
+
+	// 11 === Node.DOCUMENT_FRAGMENT_NODE
+	if (el.nodeType === 11) {
+		el = el.host; // swap to host node
+	}
+
+	if (['STYLE', 'SCRIPT'].includes(el.nodeName.toUpperCase())) {
+		return false;
+	}
+
+	const style = window.getComputedStyle(el, null);
+	if (!style) {
+		throw new Error('Style does not exist for the given element.');
+	}
+
+	if (style.getPropertyValue('display') === 'none') {
+		return true;
+	}
+
+	const visibilityValue = style.getPropertyValue('visibility');
+	if (
+		(visibilityValue === 'hidden' && !descendentVisibilityValue) ||
+		(visibilityValue === 'hidden' &&
+			(descendentVisibilityValue && descendentVisibilityValue !== 'visible'))
+	) {
+		return true;
+	}
+
+	const parent = dom.getComposedParent(el);
+	if (parent) {
+		return dom.isHiddenWithCSS(parent, visibilityValue);
```

Comment:
Nice try on the recursion, but it doesn't work this way. This function shows `#foo` as hidden, which it isn't:

```html
<div style="visibility:hidden">
  <div style="visibility:hidden">
    <div style="visibility:visible" id="foo">
      Visible
    </div>
  </div>
</div>
```

Reply:
Have tackled for the above.